zephyr/src/supp_api.c

```diff
@@ -0,0 +1,289 @@
+		/* TODO: use a timer for connection timeout*/
+		while (wpa_s->wpa_state != WPA_COMPLETED &&
+			   (ctrl->connection_timeout == SYS_FOREVER_MS ||
+			    i++ < ctrl->connection_timeout))
+		{
+			k_yield();
+			k_msleep(1000);
+		}
+
+		if (i >= ctrl->connection_timeout){
+			if (wpa_s->wpa_state != WPA_COMPLETED){
+				zephyr_supp_disconnect(ctrl->dev);
+				status = 1;
+				goto out;
+			}
+		}
+	}
+out:
+	if (ctrl->requested_op == CONNECT) {
+		wifi_mgmt_raise_connect_result_event(net_if_lookup_by_dev(ctrl->dev), status);
+	} else if (ctrl->requested_op == DISCONNECT) {
+		/* Disconnect is a synchronous operation i.e., we are already disconnected
+		 * we are just using this thread to post net_mgmt event
+		 */
+		wifi_mgmt_raise_disconnect_result_event(net_if_lookup_by_dev(ctrl->dev), 0);
+	}
+}
+
+K_THREAD_DEFINE(wpa_supp_api,
+				1024,
+				supp_shell_connect_status,
+				&wpa_supp_api_ctrl,
+				NULL,
+				NULL,
+				0,
+				0,
+				-1);
+
+
+int zephyr_supp_scan(const struct device *dev, scan_result_cb_t cb)
+{
+	const struct wifi_nrf_dev_ops *dev_ops = dev->api;
+
+	return dev_ops->off_api.disp_scan(dev, cb);
+}
+
+
+int zephyr_supp_connect(const struct device *dev,
+						struct wifi_connect_req_params *params)
+{
+	struct wpa_ssid *ssid = NULL;
+	bool pmf = true;
+	struct wpa_supplicant *wpa_s;
+
+	wpa_s = get_wpa_s_handle(dev);
+	if (!wpa_s) {
+		return -1;
+	}
+
+	wpa_supplicant_remove_all_networks(wpa_s);
+
+	ssid = wpa_supplicant_add_network(wpa_s);
+	ssid->ssid = os_zalloc(sizeof(u8) * MAX_SSID_LEN);
+
+	memcpy(ssid->ssid, params->ssid, params->ssid_length);
+	ssid->ssid_len = params->ssid_length;
+	ssid->disabled = 1;
+	ssid->key_mgmt = WPA_KEY_MGMT_NONE;
+
+	wpa_s->conf->filter_ssids = 1;
+	wpa_s->conf->ap_scan = 1;
+
+	if (params->psk) {
+		// TODO: Extend enum wifi_security_type
+		if (params->security == 3) {
+			ssid->key_mgmt = WPA_KEY_MGMT_SAE;
+			str_clear_free(ssid->sae_password);
+			ssid->sae_password = dup_binstr(params->psk, params->psk_length);
+
+			if (ssid->sae_password == NULL) {
+				wpa_printf(MSG_ERROR, "%s:Failed t copy sae_password\n",
+					      __func__);
+				return -1;
+			}
+		} else {
+			if (params->security == 2)
+				ssid->key_mgmt = WPA_KEY_MGMT_PSK_SHA256;
+			else
+				ssid->key_mgmt = WPA_KEY_MGMT_PSK;
+
+			str_clear_free(ssid->passphrase);
+			ssid->passphrase = dup_binstr(params->psk, params->psk_length);
+
+			if (ssid->passphrase == NULL) {
+				wpa_printf(MSG_ERROR, "%s:Failed t copy passphrase\n",
+				__func__);
+				return -1;
+			}
+		}
+
+		wpa_config_update_psk(ssid);
+
+		if (pmf)
+			ssid->ieee80211w = 1;
+
+	}
+
+	wpa_supplicant_enable_network(wpa_s,
+				      ssid);
+
+	wpa_supplicant_select_network(wpa_s,
+				      ssid);
+
+	send_wpa_supplicant_dummy_event();
+
+	wpa_supp_api_ctrl.dev = dev;
+	wpa_supp_api_ctrl.requested_op = CONNECT;
+	wpa_supp_api_ctrl.connection_timeout = params->timeout;
+	/* Cleanup previous spawns */
+	k_thread_abort(wpa_supp_api);
```

Comment:
Generally thread aborting in Zephyr is not considered a safe operation (https://docs.zephyrproject.org/latest/kernel/services/threads/index.html#thread-aborting), especially that the kernel gives to guarantees that it'll be able to respawn the thread. That's another reason a delayable work might be a better fit here.